java/arcs/core/policy/PolicyConstraints.kt

```diff
@@ -132,22 +133,17 @@ sealed class PolicyViolation(val policy: Policy, message: String) : Exception(
     "Policy ${policy.name} violated: $message"
 ) {
     /** Thrown when egress particles were found in the recipe that are not allowed by policy. */
-    class InvalidEgressParticles(
+    class InvalidEgressTypeForParticles(
         policy: Policy,
-        val allowedEgresses: List<String>,
-        val invalidEgresses: List<String>
+        /** Maps from particle name to egress type. */
+        val invalidEgresses: Map<String, String?>
     ) : PolicyViolation(
         policy,
-        "Invalid egress particles found: " +
-        invalidEgresses.toSortedSet().joinToString(prefix = "{", postfix = "}, ") +
-        "allowed egress particles: " +
-        allowedEgresses.toSortedSet().joinToString(prefix = "{", postfix = "}")
-    )
-
-    /** Thrown when policy has no egress particles associated with it. */
-    class PolicyHasNoEgressParticles(policy: Policy) : PolicyViolation(
-        policy,
-        "No egress particles specified for policy"
+        "Invalid egress types found for particles: ${invalidEgresses.toSortedMap().entries
+            .joinToString(prefix = "{", postfix = "}") { (name, egressType) ->
+                "$name ($egressType)"
+            }
+        }. Egress type allowed by policy: ${policy.egressType}."
```

Comment:
Might be a bit easier to read if you refactor the construction of the invalid egress types to a private function in `PolicyHasNoEgressParticles`.

Reply:
I moved some of this to a private extension function, and made the input arg simpler. WDYT?